src/Datadog.Trace.ClrProfiler.Managed/CallTarget/Handlers/Continuations/TaskContinuationGenerator.cs

```diff
@@ -47,37 +43,53 @@ public override TReturn SetContinuation(TTarget instance, TReturn returnValue, E
                 return returnValue;
             }
 
-            var continuationState = new ContinuationGeneratorState<TTarget>(instance, state);
-            return ToTReturn(previousTask.ContinueWith(
-                _continuationAction,
-                continuationState,
-                CancellationToken.None,
-                TaskContinuationOptions.ExecuteSynchronously,
-                TaskScheduler.Current));
+            return ToTReturn(ContinuationAction(previousTask, instance, state));
         }
 
-        private static void ContinuationAction(Task previousTask, object state)
+        private static async Task ContinuationAction(Task previousTask, TTarget target, CallTargetState state)
         {
-            try
+            if (!previousTask.IsCompleted)
             {
-                ContinuationGeneratorState<TTarget> contState = (ContinuationGeneratorState<TTarget>)state;
+                await new NoThrowAwaiter(previousTask);
+            }
+
+            Exception exception = null;
 
+            if (previousTask.Status == TaskStatus.Faulted)
+            {
+                exception = previousTask.Exception.GetBaseException();
+            }
+            else if (previousTask.Status == TaskStatus.Canceled)
+            {
+                try
+                {
+                    // The only supported way to extract the cancellation exception is to await the task
+                    await previousTask;
+                }
+                catch (Exception ex)
+                {
+                    exception = ex;
+                }
+            }
+
+            try
+            {
                 // *
                 // Calls the CallTarget integration continuation, exceptions here should never bubble up to the application
                 // *
-                _continuation(contState.Target, null, previousTask.Exception, contState.State);
+                _continuation(target, null, exception, state);
```

Comment:
`previousTask.Exception` wraps the exception in an AggregateException. `GetBaseException` extracts the original exception. I don't think that's something we test (it's only the callstack that is going to be shown in the span) but I believe the latter is more correct.

Reply:
So that's what I assumed, but then here we're testing for a specific exception as far as I can see:

https://github.com/DataDog/dd-trace-dotnet/blob/9a22c1653d9eb329216d26b5498eca658a8771ca/test/Datadog.Trace.Tests/CallTarget/TaskContinuationGeneratorTests.cs#L46-L47